File: tests/wasm/test_swma.js
/**
 * WASM binding tests for SWMA indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = await loadTestData();
});

test.describe('SWMA (Symmetric Weighted Moving Average)', () => {
    test('SWMA empty input', () => {
        
        const empty = new Float64Array([]);
        
        assert.throws(
            () => wasm.swma_js(empty, 5),
            /Input data slice is empty/,
            'Should throw error for empty input'
        );
    });
    
    test('SWMA accuracy', async () => {
        
        const close = new Float64Array(testData.close);
        const expected = EXPECTED_OUTPUTS.swma;
        
        const result = wasm.swma_js(
            close,
            expected.defaultParams.period
        );
        
        assert.strictEqual(result.length, close.length);
        
        
        const last5 = result.slice(-5);
        assertArrayClose(
            last5,
            expected.last5Values,
            1e-8,
            "SWMA last 5 values mismatch"
        );
        
        
        await compareWithRust('swma', result, 'close', expected.defaultParams);
    });
    
    test('basic functionality', () => {
        const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
        const period = 5;
        
        const result = wasm.swma_js(data, period);
        
        assert.strictEqual(result.length, data.length, 'Result length should match input length');
        
        
        for (let i = 0; i < period - 1; i++) {
            assert(isNaN(result[i]), `Index ${i} should be NaN during warmup`);
        }
        
        
        for (let i = period - 1; i < result.length; i++) {
            assert(!isNaN(result[i]), `Index ${i} should not be NaN after warmup`);
        }
    });
    
    test('error handling - all NaN values', () => {
        const data = new Float64Array(10).fill(NaN);
        const period = 5;
        
        assert.throws(
            () => wasm.swma_js(data, period),
            /All values are NaN/,
            'Should throw error for all NaN values'
        );
    });
    
    test('error handling - invalid period', () => {
        const data = new Float64Array([1, 2, 3, 4, 5]);
        
        
        assert.throws(
            () => wasm.swma_js(data, 6),
            /Invalid period/,
            'Should throw error when period exceeds data length'
        );
        
        
        assert.throws(
            () => wasm.swma_js(data, 0),
            /Invalid period/,
            'Should throw error for zero period'
        );
    });
    
    test('error handling - not enough valid data', () => {
        
        const data = new Float64Array([NaN, NaN, NaN, NaN, NaN, NaN, NaN, NaN, 1, 2]);
        const period = 5;
        
        assert.throws(
            () => wasm.swma_js(data, period),
            /Not enough valid data/,
            'Should throw error when not enough valid data after NaN values'
        );
    });
    
    test('leading NaN values', () => {
        const data = new Float64Array([NaN, NaN, 1, 2, 3, 4, 5, 6, 7, 8]);
        const period = 3;
        
        const result = wasm.swma_js(data, period);
        
        
        
        for (let i = 0; i < 4; i++) {
            assert(isNaN(result[i]), `Index ${i} should be NaN`);
        }
        for (let i = 4; i < result.length; i++) {
            assert(!isNaN(result[i]), `Index ${i} should not be NaN`);
        }
    });
    
    test('compare with Rust implementation', async () => {
        const close = new Float64Array(testData.close);
        const period = 5;
        
        const result = wasm.swma_js(close, period);
        await compareWithRust('swma', result, 'close', { period });
    });
    
    test('batch calculation', () => {
        const data = new Float64Array(100);
        for (let i = 0; i < 100; i++) {
            data[i] = Math.random() * 100;
        }
        
        const minPeriod = 3;
        const maxPeriod = 10;
        const stepPeriod = 2;
        
        const values = wasm.swma_batch_js(data, minPeriod, maxPeriod, stepPeriod);
        const metadata = wasm.swma_batch_metadata_js(minPeriod, maxPeriod, stepPeriod);
        
        
        const expectedPeriods = [];
        for (let p = minPeriod; p <= maxPeriod; p += stepPeriod) {
            expectedPeriods.push(p);
        }
        
        const rows = expectedPeriods.length;
        const cols = data.length;
        
        assert.strictEqual(metadata.length, rows, 'Metadata length should match number of periods');
        assert.strictEqual(values.length, rows * cols, 'Values array size should be rows*cols');
        
        
        for (let i = 0; i < expectedPeriods.length; i++) {
            const period = expectedPeriods[i];
            const individual = wasm.swma_js(data, period);
            
            
            const row = new Float64Array(cols);
            for (let j = 0; j < cols; j++) {
                row[j] = values[i * cols + j];
            }
            
            assertArrayClose(row, individual, 1e-10, `Batch row ${i} should match individual calculation`);
        }
    });
    
    test('symmetric triangular weights', () => {
        
        const data = new Float64Array(30).fill(0);
        data[15] = 1.0; 
        const period = 5;
        
        const result = wasm.swma_js(data, period);
        
        
        
        
        
        
        
        
        
        assertClose(result[15], 1/9, 1e-10, 'Weight at position 4');
        assertClose(result[16], 2/9, 1e-10, 'Weight at position 3');
        assertClose(result[17], 3/9, 1e-10, 'Weight at position 2 (center)');
        assertClose(result[18], 2/9, 1e-10, 'Weight at position 1');
        assertClose(result[19], 1/9, 1e-10, 'Weight at position 0');
        
        
        assert.strictEqual(result[10], 0.0, 'Should be zero before affected range');
        assert.strictEqual(result[20], 0.0, 'Should be zero after affected range');
    });
    
    test('edge cases', () => {
        
        const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
        const result = wasm.swma_js(data, 10);
        
        
        for (let i = 0; i < 9; i++) {
            assert(isNaN(result[i]), `Index ${i} should be NaN`);
        }
        assert(!isNaN(result[9]), 'Last value should not be NaN');
        
        
        const data2 = new Float64Array([1, 2, 3]);
        const result2 = wasm.swma_js(data2, 1);
        assertArrayClose(result2, data2, 1e-10, 'Period 1 should return input as-is');
    });
    
    test('performance with large dataset', () => {
        const data = new Float64Array(100000);
        for (let i = 0; i < data.length; i++) {
            data[i] = Math.random() * 100;
        }
        const period = 20;
        
        const start = performance.now();
        const result = wasm.swma_js(data, period);
        const elapsed = performance.now() - start;
        
        assert.strictEqual(result.length, data.length, 'Result length should match input');
        assert(elapsed < 1000, `Should process 100k elements in under 1 second, took ${elapsed}ms`);
        
        
        for (let i = 0; i < period - 1; i++) {
            assert(isNaN(result[i]), `Index ${i} should be NaN`);
        }
        assert(!isNaN(result[period - 1]), 'First non-NaN value should be at period-1');
    });
    
    
    test('SWMA zero-copy API', () => {
        const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
        const period = 5;
        
        
        const ptr = wasm.swma_alloc(data.length);
        assert(ptr !== 0, 'Failed to allocate memory');
        
        
        const memView = new Float64Array(
            wasm.__wasm.memory.buffer,
            ptr,
            data.length
        );
        
        
        memView.set(data);
        
        
        try {
            wasm.swma_into(ptr, ptr, data.length, period);
            
            
            const regularResult = wasm.swma_js(data, period);
            for (let i = 0; i < data.length; i++) {
                if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                    continue; 
                }
                assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                       `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
            }
        } finally {
            
            wasm.swma_free(ptr, data.length);
        }
    });
    
    test('SWMA zero-copy with large dataset', () => {
        const size = 100000;
        const data = new Float64Array(size);
        for (let i = 0; i < size; i++) {
            data[i] = Math.sin(i * 0.01) + Math.random() * 0.1;
        }
        
        const ptr = wasm.swma_alloc(size);
        assert(ptr !== 0, 'Failed to allocate large buffer');
        
        try {
            const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
            memView.set(data);
            
            wasm.swma_into(ptr, ptr, size, 10);
            
            
            const memView2 = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
            
            
            for (let i = 0; i < 9; i++) {
                assert(isNaN(memView2[i]), `Expected NaN at warmup index ${i}`);
            }
            
            
            for (let i = 9; i < Math.min(100, size); i++) {
                assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
            }
        } finally {
            wasm.swma_free(ptr, size);
        }
    });
    
    test('SWMA zero-copy error handling', () => {
        
        assert.throws(() => {
            wasm.swma_into(0, 0, 10, 5);
        }, /null pointer|Null pointer provided/i);
        
        
        const ptr = wasm.swma_alloc(10);
        try {
            
            assert.throws(() => {
                wasm.swma_into(ptr, ptr, 10, 0);
            }, /Invalid period/);
            
            
            assert.throws(() => {
                wasm.swma_into(ptr, ptr, 10, 11);
            }, /Invalid period/);
        } finally {
            wasm.swma_free(ptr, 10);
        }
    });
    
    test('SWMA zero-copy memory management', () => {
        
        const sizes = [100, 1000, 10000, 100000];
        
        for (const size of sizes) {
            const ptr = wasm.swma_alloc(size);
            assert(ptr !== 0, `Failed to allocate ${size} elements`);
            
            
            const memView = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
            for (let i = 0; i < Math.min(10, size); i++) {
                memView[i] = i * 1.5;
            }
            
            
            for (let i = 0; i < Math.min(10, size); i++) {
                assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
            }
            
            
            wasm.swma_free(ptr, size);
        }
    });
    
    
    test('SWMA batch - unified API with single parameter', () => {
        const close = new Float64Array(testData.close.slice(0, 100));
        
        const result = wasm.swma_batch(close, {
            period_range: [5, 5, 0]
        });
        
        
        assert(result.values, 'Should have values array');
        assert(result.combos, 'Should have combos array');
        assert(typeof result.rows === 'number', 'Should have rows count');
        assert(typeof result.cols === 'number', 'Should have cols count');
        
        
        assert.strictEqual(result.rows, 1);
        assert.strictEqual(result.cols, close.length);
        assert.strictEqual(result.combos.length, 1);
        assert.strictEqual(result.values.length, close.length);
        
        
        const combo = result.combos[0];
        assert.strictEqual(combo.period, 5);
        
        
        const oldResult = wasm.swma_js(close, 5);
        for (let i = 0; i < oldResult.length; i++) {
            if (isNaN(oldResult[i]) && isNaN(result.values[i])) {
                continue; 
            }
            assert(Math.abs(oldResult[i] - result.values[i]) < 1e-10,
                   `Value mismatch at index ${i}`);
        }
    });
    
    test('SWMA batch - unified API with multiple parameters', () => {
        const close = new Float64Array(testData.close.slice(0, 50));
        
        const result = wasm.swma_batch(close, {
            period_range: [5, 9, 2]  
        });
        
        
        assert.strictEqual(result.rows, 3);
        assert.strictEqual(result.cols, 50);
        assert.strictEqual(result.combos.length, 3);
        assert.strictEqual(result.values.length, 150);
        
        
        const expectedPeriods = [5, 7, 9];
        for (let i = 0; i < expectedPeriods.length; i++) {
            assert.strictEqual(result.combos[i].period, expectedPeriods[i]);
        }
        
        
        const firstRow = result.values.slice(0, result.cols);
        const oldResult = wasm.swma_js(close, 5);
        for (let i = 0; i < oldResult.length; i++) {
            if (isNaN(oldResult[i]) && isNaN(firstRow[i])) {
                continue; 
            }
            assert(Math.abs(oldResult[i] - firstRow[i]) < 1e-10,
                   `Value mismatch at index ${i}`);
        }
    });
    
    test('SWMA batch - unified API error handling', () => {
        const close = new Float64Array(testData.close.slice(0, 10));
        
        
        assert.throws(() => {
            wasm.swma_batch(close, {
                period_range: [5, 5]  
            });
        }, /Invalid config/);
        
        
        assert.throws(() => {
            wasm.swma_batch(close, {});
        }, /Invalid config/);
        
        
        assert.throws(() => {
            wasm.swma_batch(close, {
                period_range: "invalid"
            });
        }, /Invalid config/);
    });
    
    test('SWMA batch edge cases', () => {
        const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
        
        
        const singleBatch = wasm.swma_batch(close, {
            period_range: [5, 5, 1]
        });
        
        assert.strictEqual(singleBatch.values.length, 10);
        assert.strictEqual(singleBatch.combos.length, 1);
        
        
        const largeBatch = wasm.swma_batch(close, {
            period_range: [5, 7, 10]  
        });
        
        
        assert.strictEqual(largeBatch.values.length, 10);
        assert.strictEqual(largeBatch.combos.length, 1);
        
        
        assert.throws(() => {
            wasm.swma_batch(new Float64Array([]), {
                period_range: [5, 5, 0]
            });
        }, /Input data slice is empty/);
    });
});

File: tests/python/test_swma.py
"""
Test SWMA (Symmetric Weighted Moving Average) indicator Python bindings
"""
import numpy as np
import pytest
from my_project import swma, swma_batch, SwmaStream
from rust_comparison import compare_with_rust
from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestSwma:
    """Test SWMA indicator functionality"""

    def test_basic_functionality(self):
        """Test basic SWMA calculation"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])
        period = 5

        result = swma(data, period)

        assert isinstance(result, np.ndarray)
        assert result.shape == data.shape

        assert np.isnan(result[0:period-1]).all()
        assert not np.isnan(result[period-1:]).any()

    def test_swma_empty_input(self):
        """Test SWMA fails with empty input - mirrors check_swma_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            swma(empty, 5)

    def test_swma_accuracy(self):
        """Test SWMA matches expected values from Rust tests - mirrors check_swma_accuracy"""
        test_data = load_test_data()
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['swma']

        result = swma(
            close,
            period=expected['default_params']['period']
        )

        assert len(result) == len(close)



        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0.0,
            atol=1e-8,
            msg="SWMA last 5 values mismatch"
        )

    def test_default_period(self):
        """Test SWMA with consistent period usage"""
        data = np.random.random(100)


        result = swma(data, 5)

        assert isinstance(result, np.ndarray)
        assert result.shape == data.shape

    def test_kernel_selection(self):
        """Test different kernel selections"""
        data = np.random.random(1000)
        period = 5


        result_auto = swma(data, period, kernel='auto')
        result_scalar = swma(data, period, kernel='scalar')


        np.testing.assert_allclose(result_auto, result_scalar, rtol=1e-10)


        try:
            result_avx2 = swma(data, period, kernel='avx2')
            np.testing.assert_allclose(result_scalar, result_avx2, rtol=1e-10)
        except ValueError as e:
            msg = str(e)
            if 'not compiled' in msg or 'not available' in msg:
                pytest.skip('AVX2 kernel not available in this build/CPU')
            else:
                raise

        try:
            result_avx512 = swma(data, period, kernel='avx512')
            np.testing.assert_allclose(result_scalar, result_avx512, rtol=1e-10)
        except ValueError as e:
            msg = str(e)
            if 'not compiled' in msg or 'not available' in msg:
                pytest.skip('AVX512 kernel not available in this build/CPU')
            else:
                raise

    def test_invalid_kernel(self):
        """Test error handling for invalid kernel"""
        data = np.random.random(100)

        with pytest.raises(ValueError, match="Unknown kernel"):
            swma(data, 5, kernel='invalid_kernel')

    def test_error_all_nan(self):
        """Test error when all values are NaN"""
        data = np.full(10, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            swma(data, 5)

    def test_error_invalid_period(self):
        """Test error for invalid period values"""
        data = np.random.random(10)


        with pytest.raises(ValueError, match="Invalid period"):
            swma(data, 11)


        with pytest.raises(ValueError, match="Invalid period"):
            swma(data, 0)

    def test_error_not_enough_valid_data(self):
        """Test error when not enough valid data after NaN values"""

        data = np.array([np.nan] * 8 + [1.0, 2.0])

        with pytest.raises(ValueError, match="Not enough valid data"):
            swma(data, 5)

    def test_leading_nans(self):
        """Test SWMA with leading NaN values"""

        data = np.array([np.nan, np.nan, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0])
        period = 3

        result = swma(data, period)


        assert np.isnan(result[0:4]).all()
        assert not np.isnan(result[4:]).any()

    def test_compare_with_rust(self):
        """Test that Python bindings match Rust implementation"""
        candles = load_test_data()
        data = candles['close']
        period = 5

        result = swma(data, period)
        compare_with_rust("swma", result, 'close', params={'period': period})

    def test_swma_stream(self):
        """Test SWMA streaming functionality"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0])
        period = 5


        stream = SwmaStream(period)


        batch_result = swma(data, period)


        stream_results = []
        for value in data:
            result = stream.update(value)
            stream_results.append(result if result is not None else np.nan)

        stream_results = np.array(stream_results)


        np.testing.assert_allclose(
            batch_result[period-1:],
            stream_results[period-1:],
            rtol=1e-10
        )

    def test_swma_stream_explicit_period(self):
        """Test SWMA stream with explicit period"""

        stream = SwmaStream(5)


        data = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0]
        results = []
        for value in data:
            result = stream.update(value)
            results.append(result)


        assert results[:4] == [None, None, None, None]
        assert results[4] is not None
        assert results[5] is not None

    def test_batch_calculation(self):
        """Test batch SWMA calculation with multiple periods"""
        data = np.random.random(100)
        period_range = (3, 10, 2)

        result = swma_batch(data, period_range)

        assert isinstance(result, dict)
        assert 'values' in result
        assert 'periods' in result


        expected_periods = [3, 5, 7, 9]
        assert result['values'].shape == (len(expected_periods), len(data))
        np.testing.assert_array_equal(result['periods'], expected_periods)


        for i, period in enumerate(expected_periods):
            individual_result = swma(data, period)
            np.testing.assert_allclose(
                result['values'][i],
                individual_result,
                rtol=1e-10
            )

    def test_batch_kernel_selection(self):
        """Test batch calculation with different kernels"""
        data = np.random.random(100)
        period_range = (5, 10, 5)


        result_auto = swma_batch(data, period_range, kernel='auto')
        result_scalar = swma_batch(data, period_range, kernel='scalar')

        np.testing.assert_allclose(result_auto['values'], result_scalar['values'], rtol=1e-10)

        try:
            result_avx2 = swma_batch(data, period_range, kernel='avx2')
            np.testing.assert_allclose(result_scalar['values'], result_avx2['values'], rtol=1e-10)
        except ValueError as e:
            msg = str(e)
            if 'not compiled' in msg or 'not available' in msg:
                pytest.skip('AVX2 batch kernel not available in this build/CPU')
            else:
                raise

        try:
            result_avx512 = swma_batch(data, period_range, kernel='avx512')
            np.testing.assert_allclose(result_scalar['values'], result_avx512['values'], rtol=1e-10)
        except ValueError as e:
            msg = str(e)
            if 'not compiled' in msg or 'not available' in msg:
                pytest.skip('AVX512 batch kernel not available in this build/CPU')
            else:
                raise

    def test_batch_invalid_kernel(self):
        """Test batch calculation with invalid kernel"""
        data = np.random.random(100)
        period_range = (5, 10, 5)

        with pytest.raises(ValueError, match="Unknown kernel"):
            swma_batch(data, period_range, kernel='invalid')

    def test_zero_copy_performance(self):
        """Test that zero-copy operations are working (indirect test)"""

        data = np.random.random(1_000_000)
        period = 20


        result = swma(data, period)


        assert result.shape == data.shape
        assert np.isnan(result[:period-1]).all()
        assert not np.isnan(result[period-1:]).any()

    def test_symmetric_weights(self):
        """Test that SWMA uses symmetric triangular weights"""

        data = np.zeros(30)
        data[15] = 1.0
        period = 5

        result = swma(data, period)

















        assert abs(result[15] - 1/9) < 1e-10
        assert abs(result[16] - 2/9) < 1e-10
        assert abs(result[17] - 3/9) < 1e-10
        assert abs(result[18] - 2/9) < 1e-10
        assert abs(result[19] - 1/9) < 1e-10


        assert result[10] == 0.0
        assert result[20] == 0.0

    def test_real_world_data(self):
        """Test with real-world conditions including warmup period"""

        np.random.seed(42)
        trend = np.linspace(100, 110, 200)
        noise = np.random.normal(0, 0.5, 200)
        data = trend + noise

        period = 10
        result = swma(data, period)


        assert np.isnan(result[:period-1]).all()
        assert not np.isnan(result[period-1:]).any()


        input_volatility = np.std(np.diff(data[period:]))
        output_volatility = np.std(np.diff(result[period:]))
        assert output_volatility < input_volatility

    def test_edge_cases(self):
        """Test edge cases"""

        data = np.array([42.0])
        with pytest.raises(ValueError, match="Invalid period"):
            swma(data, 2)


        data = np.random.random(10)
        result = swma(data, 10)
        assert np.isnan(result[:-1]).all()
        assert not np.isnan(result[-1])


        data = np.array([1.0, 2.0, 3.0])
        result = swma(data, 1)
        np.testing.assert_array_equal(result, data)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
